features/nvstore/source/nvstore.cpp

```diff
@@ -0,0 +1,1066 @@
+static uint32_t crc32(uint32_t init_crc, uint32_t data_len, uint8_t *data_buf)
+{
+    uint32_t i, j;
+    uint32_t crc, mask;
+
+    crc = init_crc;
+    for (i = 0; i < data_len; i++) {
+       crc = crc ^ (uint32_t) (data_buf[i]);
+       for (j = 0; j < 8; j++) {
+          mask = -(crc & 1);
+          crc = (crc >> 1) ^ (0xEDB88320 & mask);
+       }
+    }
+    return crc;
+}
+
+NVStore::NVStore() : _init_done(0), _init_attempts(0), _active_area(0), _max_keys(NVSTORE_MAX_KEYS),
+               _active_area_version(0), _free_space_offset(0), _size(0), _offset_by_key(0)
+{
+}
+
+NVStore::~NVStore()
+{
+    if (_init_done) {
+        deinit();
+    }
+}
+
+uint16_t NVStore::get_max_keys() const
+{
+    return _max_keys;
+}
+
+void NVStore::set_max_keys(uint16_t num_keys)
+{
+    MBED_ASSERT(num_keys < MASTER_RECORD_KEY);
+    _max_keys = num_keys;
+    // User is allowed to change number of types. As this affects init, need to deinitialize now.
+    // Don't call init right away - it is lazily called by get/set functions if needed.
+    deinit();
+}
+
+// Flash access helper functions, using area and offset notations
+
+// Read from flash, given area and offset.
+// Parameters :
+// area          - [IN]   Flash area.
+// offset        - [IN]   Offset in area.
+// len_bytes     - [IN]   Length in bytes.
+// buf           - [IN]   Data buffer.
+// Return        : 0 on success. Error code otherwise.
+int NVStore::flash_read_area(uint8_t area, uint32_t offset, uint32_t len_bytes, uint32_t *buf)
+{
+    return nvstore_int_flash_read(len_bytes, _flash_area_params[area].address + offset, buf);
+}
+
+// Write to flash, given area and offset.
+// Parameters :
+// area          - [IN]   Flash area.
+// offset        - [IN]   Offset in area.
+// len_bytes     - [IN]   Length in bytes.
+// buf           - [IN]   Data buffer.
+// Return        : 0 on success. Error code otherwise.
+int NVStore::flash_write_area(uint8_t area, uint32_t offset, uint32_t len_bytes, const uint32_t *buf)
+{
+    return nvstore_int_flash_write(len_bytes, _flash_area_params[area].address + offset, buf);
+}
+
+// Erase a flash area, given area.
+// Parameters :
+// area          - [IN]   Flash area.
+// Return        : 0 on success. Error code otherwise.
+int NVStore::flash_erase_area(uint8_t area)
+{
+    return nvstore_int_flash_erase(_flash_area_params[area].address, _flash_area_params[area].size);
+}
+
+// Calculate area address and size configuration (in case it's not determined by the user)
+// or validate it (in case it is)
+void NVStore::calc_validate_area_params()
+{
+    int num_sectors = 0;
+
+    size_t flash_addr = nvstore_int_flash_get_flash_start();
+    size_t flash_size = nvstore_int_flash_get_flash_size();
+    size_t sector_size;
+    int max_sectors = flash_size / nvstore_int_flash_get_sector_size(flash_addr) + 1;
+    size_t *sector_map = new size_t[max_sectors];
+
+    int area = 0;
+    size_t left_size = flash_size;
+#if AREA_PARAMS_USER_CONFIG
```

Comment:
Why not use a ternary to assign the address and size? I dont understand why there would be a difference between user specified and default sectors/sizes other than the location?!?

Reply:
User can decide to locate NVStore anywhere in the flash, for instance in the beginning in components that have small sectors. In this case there would be a difference in both addresses and sizes of the areas. User can decide to use a larger area (and use multiple consecutive sectors) knowing that NVStore should store a lot of data.
Not sure I understand how ternary would help in initialization of this array. Please elaborate.

Reply:
So the logic handles single sector per block or multiple sectors per block? Regardless of where the address and size come from, why is this a compile time conditional block of code rather than the same implementation that operated on the address and sizes specified?

Reply:
OK, seems I didn't explain this properly, so let me elaborate:
First of all, let's use NVStore terminology, which is areas (instead of blocks). Each area consists of one or more consecutive sectors. Area addresses and sizes get their values once during initialization, which is when this function is called. Once set, sectors aren't of our concern any more, we just use area addresses and sizes (hence the required change in FlashIAP programming).
Now, areas can get their addresses and sizes in one of two ways - from the user (via configuration, supplying us address and size defines per area) or by automatic calculation, in the case these defines are missing from the configuration, which is the default. The only way for NVStore to tell which case it is, is according to these defines it gets (or whether they are missing). NVStore should behave differently on each case. In the user case it should just validate the user configuration (based on the sector map). In the other case (let's call it non-user case), it takes the last two sectors (or more, depending on minimum size) for this purpose, also based on sector map. Both cases are distinguished by these defines, imposed on NVStore, and have different logics. I could have separated the cases to different functions, but both would still need to use the sector map, which would create a redundant overhead. I could use a run-time flag based on these defines, but I don't think it is the way to go. Can do, if the site of code that's separated by defines looks unpleasant. 

Just to demonstrate a real life scenario: Some boards (like NUCLO boards) have large sectors as their last ones. Right now we don't have tools that move the program forward to use the first, smaller sectors, so by default we will use the non-user case. This case will trigger the non-user code at startup (based on the fact we don't have defines for area addresses and sizes). Once these tools are implemented, NVStore can get the addresses and sizes by defines, and act differently at startup.

Reply:
> Now, areas can get their addresses and sizes in one of two ways - from the user (via configuration, supplying us address and size defines per area) or by automatic calculation, in the case these defines are missing from the configuration, which is the default

Why don't both methods result in the use of the same define? So if not user config, use default config both of which would result in the same MACROS being populated for the area address and area size.

Reply:
Having said that, I agree that the multiple usage of `#ifdef` in this area of code is not a pleasant site. Will refactor the code to make it look better.

Reply:
@sg- Refactored the code. Please have a look.